spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfigurationIntegrationTests.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * @author Erik Merkle
+ */
+@Testcontainers(disabledWithoutDocker = true)
+public class CassandraMetricsAutoConfigurationIntegrationTests {
+
+	@Container
+	static final CassandraContainer<?> cassandra = new CassandraContainer<>().withStartupAttempts(5)
+			.withStartupTimeout(Duration.ofMinutes(10)).waitingFor(new CassandraWaitStrategy());
+
+	private final ApplicationContextRunner contextRunner = new ApplicationContextRunner().with(MetricsRun.simple())
+			.withConfiguration(
+					AutoConfigurations.of(CassandraMetricsAutoConfiguration.class, CassandraAutoConfiguration.class))
+			.withPropertyValues(
+					"spring.data.cassandra.contact-points:" + cassandra.getHost() + ":"
+							+ cassandra.getFirstMappedPort(),
+					"spring.data.cassandra.local-datacenter=datacenter1", "spring.data.cassandra.read-timeout=20s",
+					"spring.data.cassandra.connect-timeout=10s");
+
+	/**
+	 * Cassandra driver metrics should be enabled by default as long as the desired
+	 * metrics are enabled in the Driver's configuration.
+	 */
+	@Test
+	void autoConfiguredCassandraIsInstrumented() {
+		this.contextRunner.withUserConfiguration(SimpleDriverConfigLoaderBuilderMetricsConfig.class).run((context) -> {
+			CqlSessionBuilder builder = context.getBean(CqlSessionBuilder.class);
+			assertThat(builder).isInstanceOf(CassandraMetricsPostProcessor.MetricsSessionBuilder.class);
+			MeterRegistry registry = context.getBean(MeterRegistry.class);
+			// execute queries to peg metrics
+			CqlSession session = context.getBean(CqlSession.class);
+			SimpleStatement statement = SimpleStatement.newInstance("SELECT release_version FROM system.local")
+					.setConsistencyLevel(ConsistencyLevel.LOCAL_ONE);
+			for (int i = 0; i < 10; ++i) {
+				assertThat(session.execute(statement).one()).isNotNull();
+			}
+			// assert Session metrics
+			String sessionMetricPrefix = session.getContext().getSessionName() + ".";
+			assertThat(
+					registry.get(sessionMetricPrefix + DefaultSessionMetric.CONNECTED_NODES.getPath()).gauge().value())
+							.isEqualTo(1d);
+			assertThat(registry.get(sessionMetricPrefix + DefaultSessionMetric.CQL_REQUESTS.getPath()).timer().count())
+					.isEqualTo(10L);
+			assertThat(registry.get(sessionMetricPrefix + DefaultSessionMetric.BYTES_SENT.getPath()).counter().count())
+					.isGreaterThan(1d);
+			assertThat(
+					registry.get(sessionMetricPrefix + DefaultSessionMetric.BYTES_RECEIVED.getPath()).counter().count())
+							.isGreaterThan(1d);
+			// assert Node metrics
+			String nodeMetricPrefix = sessionMetricPrefix + "nodes." + cassandra.getHost() + ":"
+					+ cassandra.getMappedPort(9042) + ".";
+			assertThat(registry.get(nodeMetricPrefix + DefaultNodeMetric.BYTES_SENT.getPath()).counter().count())
+					.isGreaterThan(1d);
+
+		});
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	static class SimpleDriverConfigLoaderBuilderMetricsConfig {
+
+		@Bean
+		DriverConfigLoaderBuilderCustomizer customizer() {
```

Comment:
If that represents what users have to do in order to get metrics support, I think something is missing.

Reply:
For Driver 4.x, metrics are not enabled by default. To enable metrics in the driver, the driver must be configured with the list of enabled metrics. This can be done programmatically, with System properties or by providing a driver config file. For this test, it was easiest to just provide a Driver config customizer bean. I can change the test to enable the metrics under test via System properties or with an external driver configuration file if you prefer.

Are you concerned with how metrics are enabled in the driver? Or should I be enabling metrics automatically in the Actuator?

Reply:
> Are you concerned with how metrics are enabled in the driver? Or should I be enabling metrics automatically in the Actuator?

If we provide first-class support for metrics here, the user shouldn't have to write code to enable it ideally. Looking at other metrics we have, they are usually enabled by default, sometimes with a flag that indicates if metrics for that particular `CqlSession` is enabled or not. Concretely we should then have a `enabled` property somewhere in the `spring.cassandra` namespace that user can set in `application.properties` the usual way.

I don't have an opinion as whether the flag should be enabled by default or not but listing the metrics seem a bit tedious and inconsistent for an "out-of-the-box" scenario.

Paging @shakuzen to get some more feedback on this.

Reply:
I'm not very familiar with Cassandra driver specifics, but it's probably only metrics on specific queries which might have high cardinality or privacy concerns (if query parameters/values are tagged, for example) that we wouldn't want to enable by default, I think. Connection pool or node or session metrics seem generally useful and probably safe from the aforementioned concerns.